### ISSUE TITLE
`TerminateOnNaN` callback triggers `on_train_end()`, causing unintended side effects in other callbacks

### ISSUE THREAD

**Issue by PhilLord0000** [keras-team/keras#21771]
The `TerminateOnNaN` callback currently terminates training by setting `self.model.stop_training = True` when NaN or Inf loss is detected. Whilst this gracefully stops the training loop, it also triggers the `on_train_end()` method for all other callbacks, which can have unintended negative consequences.
Examples of problematic interactions:
`BackupAndRestore`: Deletes its backup directory in `on_train_end()`, preventing recovery from the last good epoch when NaN occurs
`EarlyStopping`: May restore best weights based on an incomplete training run

There might be more then that, I initially ran into this issue when using `BackupAndRestore`. I've been using this custom callback to stop the training without running the other callbacks.

```
class HardTerminateOnNaN(tf.keras.callbacks.Callback):
    def on_batch_end(self, batch, logs=None):
        loss = logs.get("loss")
        if loss is not None and (tf.math.is_nan(loss) or tf.math.is_inf(loss)):
            print(f"\nNaN detected at batch {batch}. Terminating immediately.")
            raise RuntimeError("NaN loss encountered.")
```

I propose either adding a new callback, similar to the one above or adding an option to the existing `TerminateOnNaN`. I'd be happy to open a PR and implement this based on what you think would be the better option.

**Comment by dhantule** [comment]
Hi @PhilLord0000, thanks for reporting this.

Can you provide standalone code to reproduce this issue ?

**Comment by PhilLord0000** [comment]
Hi @dhantule, sorry that I've take a while to get back to you, I've been very busy recently.

I've created a repo with two examples, one to reproduce the issue and one with an example of my proposed callback. 
https://github.com/PhilLord0000/TerminateOnNaN_Fix

Both examples use a normal learning rate for the first epoch and then increase the learning rate to a very high value to instigate a NaN, in `reproduce_issue` you will notice that there is no backup folder after the training is terminated however the folder persists after the termination in `fix_issue`. I hope these examples provide a clear demonstration of the issue.

**Comment by MalyalaKarthik66** [comment]
Hi, @PhilLord0000 @dhantule 
I created this PR to address the issue where TerminateOnNaN only stops training gracefully, causing other callbacks like BackupAndRestore to run even after a NaN loss.

This PR introduces a new callback — HardTerminateOnNaN — that immediately raises a RuntimeError when NaN or Inf losses are detected.
This ensures training is forcefully stopped and subsequent callbacks are not triggered, matching expected behavior during catastrophic training failures.

All tests have been added and verified locally.

**Comment by MalyalaKarthik66** [comment]
@PhilLord0000 
Could you please clarify what’s incorrect or missing in this PR so I can fix it?

**Comment by PhilLord0000** [comment]
@MalyalaKarthik66 As I said in my description there are two ways that this could be implemented, either adding a new callback, or adding an option to the existing TerminateOnNaN. Your implementation was a half-way between both, @fchollet has indicated in your PR that the latter option is the preferred solution. This kind of discussion about implementation details is the point of opening an issue before creating a PR. It is bad form to take someone else's code and open a PR when they have indicated that they are going to open a PR and just want feedback from the maintainers first, especially before this discussion has even taken place.

**Comment by MalyalaKarthik66** [comment]
@PhilLord0000 
Sorry about that — I didn’t mean to step on your contribution. I saw the issue and thought it was open for implementation. My intention was only to contribute, not to take over your work.
If you prefer, I’m happy to close my PR and let you proceed with your planned implementation.

**Comment by PhilLord0000** [comment]
@MalyalaKarthik66 There's not point closing your PR, just keep this in mind for the future.

**Comment by MalyalaKarthik66** [comment]
@PhilLord0000 
Thanks for understanding. I’m really sorry about the confusion — I’ll definitely keep this in mind for future contributions.
